ckan/tests/pytest_ckan/fixtures.py

```diff
@@ -51,61 +66,115 @@
+# can define module-local version of ``user_factory`` or use parametrization.
+#
+# Because ``user_factory`` is just an alias for the factory class, you still
+# can use all the features of factory, as:
+# * custom properties: user_factory(name="Test", password="123")
+# * get model instead of a dict: user_factory.model(name="Tes") # model.User
+# * batch generation: user_factory.create_batch(100) # create 100 users
+#
+#
+# # user  - instance-fixture
+#
+# This is a shortcut for creating an instance using factory. Roughly, this::
+#
+#     def test_something(user_factory):
+#         user = user_factory()
+#         ...
+#
+# is identical to this::
+#
+#     def test_something(user):
+#         ...
+#
+# You just save one line and 8 characters(``_factory`` suffix). But apart from
+# that it's still a pytest fixture. It means that all cool effects of fixtures
+# is applicable to it. For example, if there is a fixture that depends on
+# ``user``::
+#
+#     @pytest.fixture()
+#     def user_from_somewhere(user):
+#         return user
+#
+# and a test that requires both ``user`` and this fixture from above, they will
+# share the same ``user``::
+#
+#     def test_identity(user, user_from_somewhere):
+#         assert user is user_from_somewhere
+#
+#
+# Details about factories and reference to pytest-factoryboy can be found in
+# ckan.test.factories module
+for (name, factory) in inspect.getmembers(factories, _is_factory_class):
+    # Add `Factory` suffix to the factory class name. Class name will be
+    # converted into a name for factory-fixture. We want it to be different
+    # from the name of the instance-fixture. For end-user it will be more
+    # clear, if factory-fixture use name `something_factory`, while instance
+    # fixtures will use just `something`.
+    #
+    # If we use original class name, HelloWorld class will turn into
+    # `hello_world` factory-fixture, that will be **an alias** for HelloWorld
+    # class. It means, we won't be able to use `hello_world` identifier as and
+    # instance name. We'll have to come up with something different, like
+    # `hello_world_instance`. It's not so cool, because people want shorter
+    # names for instances.
+    #
+    # That's why we are using `Factory` suffix. It will give us
+    # `hello_world_factory` fixture and we'll be able to use `hello_world` as a
+    # name for instance.
+    if not name.endswith("Factory"):
+        name += "Factory"
+
+    # Create a new class with the "right" name(suffixed by `Factory`).
+    cls = type(name, (factory, ), {})
+
+    # this function creates both fixtures
+    register(
+        # name of the factory-fixture will be obtained from `cls.__name__`
+        cls,
+        # name of instance-fixture must be passed explicitly. Strip `Factory`
+        # suffix from instance name and convert it to underscore-form.
+        inflection.underscore(name[:-7])
+    )
```

Comment:
I don't love the ruby-style automatic camel case to underscore name generation. Also the comments say that all CKANFactory subclasses will be registered this way but the implementation only applies to classes in this one module.

Can we lose the magic and use a class decorator with the names we want registered instead? This could be used equally by extensions and is easily grep-able in the source code, e.g.

```python
@register_fixtures('dataset', 'dataset_factory')
class Dataset(CKANFactory):
    """A factory class for creating CKAN datasets."""
```